{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Pipeline for variant caling in *Mtb***\n",
    "\n",
    "This Jupyter Notebook has been developed by Rita Nóbrega-Martins (PG46733) for the curicular unit 'Project' of the Master's in Bioinformatics, University of Minho. \n",
    "This work has been supervised by Dr. Nuno Osório (ICVS, University of Minho) and Dr. Tiago Beites (i3S, University of Porto).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This pipeline runs on Linux system or WSL-Ubuntu that runs a conda environment named 'MTB_calling'.\n",
    "This conda environment contains the following tools:\n",
    "* `bcftools`              version   1.11               \n",
    "* `bedtools`              version   2.30.0 \n",
    "* `bwa`                   version   0.7.17\n",
    "* `bzip2`                 version   1.0.8 \n",
    "* `fastqc`                version   0.12.1 \n",
    "* `htslib`                version   1.19.1\n",
    "* `ipykernel`             version   6.29.3           \n",
    "* `ipython`               version   8.22.2\n",
    "* `lofreq`                version   2.1.5\n",
    "* `mosdepth`              version   0.3.3 \n",
    "* `multiqc`               version   1.21 \n",
    "* `parallel`              version   20160622\n",
    "* `pathogen-profiler`     version   2.0.1 \n",
    "* `samtools`              version   1.12\n",
    "* `sra-tools`             version   3.1.0\n",
    "* `tabixpp`               version   1.1.2               \n",
    "* `tb-profiler`           version   4.3.0 \n",
    "* `trimmomatic`           version   0.39\n",
    "* `vcflib`                version   1.0.9\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "conda activate MTB_calling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **1.FASTQ extraction**\n",
    "\n",
    "**Input:** Text file with a list of SRA accession numbers of the *Mtb* isolates\n",
    "\n",
    "The funcion `extractSRAfiles()` iterates each line of the text file cointaining the SRA number of a given isolate, creating a directory for each isolate/sample, where information and files regarding that specific isolate will be stored.\n",
    "\n",
    "Through the tools `prefetch` and `fasterq-dump`, SRA data will be downnloaded from the NCBI database, and converted to FastQ files [1].\n",
    "\n",
    "**Output:** 2 FastQ files per sample\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "metadata": {},
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "extractSRAfiles () {\n",
    "    echo \"Extracting FastQ files\"\n",
    "    local sample=\"$1\" \n",
    "    local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "\n",
    "    while IFS= read -r sample; do\n",
    "        echo -e \"\\nSample: $sample\"\n",
    "        mkdir -p \"$sra_diretoria/$sample\"  \n",
    "        cd \"$sra_diretoria/$sample\" || exit 1 \n",
    "\n",
    "        prefetch \"$sample\" \n",
    "        fasterq-dump \"$sample\"\n",
    "        rm $HOME/Projeto_Mtb/SRA_files/sra/\"$sample\".sra\n",
    "    \n",
    "    done < \"$sample\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "metadata": {},
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting FastQ files\n",
      "\n",
      "Sample: DRR034421\n",
      "\n",
      "2024-06-14T14:02:52 prefetch.3.1.0: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.\n",
      "2024-06-14T14:02:52 prefetch.3.1.0: 1) 'DRR034421' is found locally\n",
      "2024-06-14T14:02:52 prefetch.3.1.0: 'DRR034421' has 0 unresolved dependencies\n",
      "spots read      : 606,259\n",
      "reads read      : 1,212,518\n",
      "reads written   : 1,212,518\n",
      "\n",
      "Sample: DRR130077\n",
      "\n",
      "2024-06-14T14:03:10 prefetch.3.1.0: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.\n",
      "2024-06-14T14:03:10 prefetch.3.1.0: 1) 'DRR130077' is found locally\n",
      "2024-06-14T14:03:11 prefetch.3.1.0: 'DRR130077' has 0 unresolved dependencies\n",
      "spots read      : 573,970\n",
      "reads read      : 1,147,940\n",
      "reads written   : 1,147,940\n",
      "\n",
      "Sample: DRR130078\n",
      "\n",
      "2024-06-14T14:03:31 prefetch.3.1.0: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.\n",
      "2024-06-14T14:03:32 prefetch.3.1.0: 1) 'DRR130078' is found locally\n",
      "2024-06-14T14:03:32 prefetch.3.1.0: 'DRR130078' has 0 unresolved dependencies\n",
      "spots read      : 863,537\n",
      "reads read      : 1,727,074\n",
      "reads written   : 1,727,074\n",
      "\n",
      "Sample: DRR130093\n",
      "\n",
      "2024-06-14T14:04:10 prefetch.3.1.0: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.\n",
      "2024-06-14T14:04:10 prefetch.3.1.0: 1) 'DRR130093' is found locally\n",
      "2024-06-14T14:04:10 prefetch.3.1.0: 'DRR130093' has 0 unresolved dependencies\n",
      "spots read      : 1,368,077\n",
      "reads read      : 2,736,154\n",
      "reads written   : 2,736,154\n"
     ]
    }
   ],
   "source": [
    "extractSRAfiles ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **2. Reads Pre-processing**\n",
    "\n",
    "The pre-processing of the reads present in the FASTQ-files extracted consists on extraction of the FASTQC file, which contain a Quality Report, and the trimming of the reads according to its quality. \n",
    "\n",
    "The function `quality_control` receives the text file with the list of isolates and iterates through each line/isolate to access the sample folder and through `fastqc` download the FASTQC-file (in a html format) per FASTQ-file, with relevant information for the assessment of the quality of the reads.\n",
    "\n",
    "The function `trimming_bbduk` iterates through the lines of the text file given as an input to access the 2 FASTQ-files of each isolate/sample and perform a trimming of the reads through the `BBDuk` tool [2] to remove adapters and low-quality reads, as well as reads that may be too short for further analysis.\n",
    "`trimming_bbduk` returns the 2 FASTQ-files with the trimmed reads per sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "metadata": {},
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "quality_control () {\n",
    "  echo 'Quality control'\n",
    "  local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "  local sample_file=\"$1\"\n",
    "  sed -i 's/\\r$//' \"$sample_file\"  \n",
    "\n",
    "  if [ ! -f \"$sample_file\" ]; then\n",
    "      echo \"Sample '$sample_file' file not found.\"\n",
    "      return 1\n",
    "  fi\n",
    "\n",
    "  parallel -j \"$(nproc)\" \"\n",
    "      sample={};\n",
    "      if [ -f \\\"$sra_diretoria/\\$sample/\\${sample}_1.fastq\\\" ] && [ -f \\\"$sra_diretoria/\\$sample/\\${sample}_2.fastq\\\" ]; then\n",
    "          fastqc \\\"$sra_diretoria/\\$sample/\\${sample}_1.fastq\\\" \\\"$sra_diretoria/\\$sample/\\${sample}_2.fastq\\\";\n",
    "      else\n",
    "          echo \\\"FASTQ not found for: '\\$sample'.\\\";\n",
    "      fi\n",
    "    \" < \"$sample_file\"\n",
    "}\n",
    "\n",
    "    # vizualização da qualidade - basta o report html ou uso o Qualitymap ou o fastxtoolkit (permitem ver na command line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "metadata": {},
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Quality control\n",
      "Academic tradition requires you to cite works you base your article on.\n",
      "When using programs that use GNU Parallel to process data for publication\n",
      "please cite:\n",
      "\n",
      "  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,\n",
      "  ;login: The USENIX Magazine, February 2011:42-47.\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "To silence the citation notice: run 'parallel --citation'.\n",
      "\n",
      "null\n",
      "null\n",
      "Analysis complete for DRR034421_1.fastq\n",
      "Analysis complete for DRR034421_2.fastq\n",
      "Started analysis of DRR034421_1.fastq\n",
      "Approx 5% complete for DRR034421_1.fastq\n",
      "Approx 10% complete for DRR034421_1.fastq\n",
      "Approx 15% complete for DRR034421_1.fastq\n",
      "Approx 20% complete for DRR034421_1.fastq\n",
      "Approx 25% complete for DRR034421_1.fastq\n",
      "Approx 30% complete for DRR034421_1.fastq\n",
      "Approx 35% complete for DRR034421_1.fastq\n",
      "Approx 40% complete for DRR034421_1.fastq\n",
      "Approx 45% complete for DRR034421_1.fastq\n",
      "Approx 50% complete for DRR034421_1.fastq\n",
      "Approx 55% complete for DRR034421_1.fastq\n",
      "Approx 60% complete for DRR034421_1.fastq\n",
      "Approx 65% complete for DRR034421_1.fastq\n",
      "Approx 70% complete for DRR034421_1.fastq\n",
      "Approx 75% complete for DRR034421_1.fastq\n",
      "Approx 80% complete for DRR034421_1.fastq\n",
      "Approx 85% complete for DRR034421_1.fastq\n",
      "Approx 90% complete for DRR034421_1.fastq\n",
      "Approx 95% complete for DRR034421_1.fastq\n",
      "Started analysis of DRR034421_2.fastq\n",
      "Approx 5% complete for DRR034421_2.fastq\n",
      "Approx 10% complete for DRR034421_2.fastq\n",
      "Approx 15% complete for DRR034421_2.fastq\n",
      "Approx 20% complete for DRR034421_2.fastq\n",
      "Approx 25% complete for DRR034421_2.fastq\n",
      "Approx 30% complete for DRR034421_2.fastq\n",
      "Approx 35% complete for DRR034421_2.fastq\n",
      "Approx 40% complete for DRR034421_2.fastq\n",
      "Approx 45% complete for DRR034421_2.fastq\n",
      "Approx 50% complete for DRR034421_2.fastq\n",
      "Approx 55% complete for DRR034421_2.fastq\n",
      "Approx 60% complete for DRR034421_2.fastq\n",
      "Approx 65% complete for DRR034421_2.fastq\n",
      "Approx 70% complete for DRR034421_2.fastq\n",
      "Approx 75% complete for DRR034421_2.fastq\n",
      "Approx 80% complete for DRR034421_2.fastq\n",
      "Approx 85% complete for DRR034421_2.fastq\n",
      "Approx 90% complete for DRR034421_2.fastq\n",
      "Approx 95% complete for DRR034421_2.fastq\n",
      "null\n",
      "null\n",
      "Analysis complete for DRR130077_1.fastq\n",
      "Analysis complete for DRR130077_2.fastq\n",
      "Started analysis of DRR130077_1.fastq\n",
      "Approx 5% complete for DRR130077_1.fastq\n",
      "Approx 10% complete for DRR130077_1.fastq\n",
      "Approx 15% complete for DRR130077_1.fastq\n",
      "Approx 20% complete for DRR130077_1.fastq\n",
      "Approx 25% complete for DRR130077_1.fastq\n",
      "Approx 30% complete for DRR130077_1.fastq\n",
      "Approx 35% complete for DRR130077_1.fastq\n",
      "Approx 40% complete for DRR130077_1.fastq\n",
      "Approx 45% complete for DRR130077_1.fastq\n",
      "Approx 50% complete for DRR130077_1.fastq\n",
      "Approx 55% complete for DRR130077_1.fastq\n",
      "Approx 60% complete for DRR130077_1.fastq\n",
      "Approx 65% complete for DRR130077_1.fastq\n",
      "Approx 70% complete for DRR130077_1.fastq\n",
      "Approx 75% complete for DRR130077_1.fastq\n",
      "Approx 80% complete for DRR130077_1.fastq\n",
      "Approx 85% complete for DRR130077_1.fastq\n",
      "Approx 90% complete for DRR130077_1.fastq\n",
      "Approx 95% complete for DRR130077_1.fastq\n",
      "Started analysis of DRR130077_2.fastq\n",
      "Approx 5% complete for DRR130077_2.fastq\n",
      "Approx 10% complete for DRR130077_2.fastq\n",
      "Approx 15% complete for DRR130077_2.fastq\n",
      "Approx 20% complete for DRR130077_2.fastq\n",
      "Approx 25% complete for DRR130077_2.fastq\n",
      "Approx 30% complete for DRR130077_2.fastq\n",
      "Approx 35% complete for DRR130077_2.fastq\n",
      "Approx 40% complete for DRR130077_2.fastq\n",
      "Approx 45% complete for DRR130077_2.fastq\n",
      "Approx 50% complete for DRR130077_2.fastq\n",
      "Approx 55% complete for DRR130077_2.fastq\n",
      "Approx 60% complete for DRR130077_2.fastq\n",
      "Approx 65% complete for DRR130077_2.fastq\n",
      "Approx 70% complete for DRR130077_2.fastq\n",
      "Approx 75% complete for DRR130077_2.fastq\n",
      "Approx 80% complete for DRR130077_2.fastq\n",
      "Approx 85% complete for DRR130077_2.fastq\n",
      "Approx 90% complete for DRR130077_2.fastq\n",
      "Approx 95% complete for DRR130077_2.fastq\n",
      "null\n",
      "null\n",
      "Analysis complete for DRR130078_1.fastq\n",
      "Analysis complete for DRR130078_2.fastq\n",
      "Started analysis of DRR130078_1.fastq\n",
      "Approx 5% complete for DRR130078_1.fastq\n",
      "Approx 10% complete for DRR130078_1.fastq\n",
      "Approx 15% complete for DRR130078_1.fastq\n",
      "Approx 20% complete for DRR130078_1.fastq\n",
      "Approx 25% complete for DRR130078_1.fastq\n",
      "Approx 30% complete for DRR130078_1.fastq\n",
      "Approx 35% complete for DRR130078_1.fastq\n",
      "Approx 40% complete for DRR130078_1.fastq\n",
      "Approx 45% complete for DRR130078_1.fastq\n",
      "Approx 50% complete for DRR130078_1.fastq\n",
      "Approx 55% complete for DRR130078_1.fastq\n",
      "Approx 60% complete for DRR130078_1.fastq\n",
      "Approx 65% complete for DRR130078_1.fastq\n",
      "Approx 70% complete for DRR130078_1.fastq\n",
      "Approx 75% complete for DRR130078_1.fastq\n",
      "Approx 80% complete for DRR130078_1.fastq\n",
      "Approx 85% complete for DRR130078_1.fastq\n",
      "Approx 90% complete for DRR130078_1.fastq\n",
      "Approx 95% complete for DRR130078_1.fastq\n",
      "Started analysis of DRR130078_2.fastq\n",
      "Approx 5% complete for DRR130078_2.fastq\n",
      "Approx 10% complete for DRR130078_2.fastq\n",
      "Approx 15% complete for DRR130078_2.fastq\n",
      "Approx 20% complete for DRR130078_2.fastq\n",
      "Approx 25% complete for DRR130078_2.fastq\n",
      "Approx 30% complete for DRR130078_2.fastq\n",
      "Approx 35% complete for DRR130078_2.fastq\n",
      "Approx 40% complete for DRR130078_2.fastq\n",
      "Approx 45% complete for DRR130078_2.fastq\n",
      "Approx 50% complete for DRR130078_2.fastq\n",
      "Approx 55% complete for DRR130078_2.fastq\n",
      "Approx 60% complete for DRR130078_2.fastq\n",
      "Approx 65% complete for DRR130078_2.fastq\n",
      "Approx 70% complete for DRR130078_2.fastq\n",
      "Approx 75% complete for DRR130078_2.fastq\n",
      "Approx 80% complete for DRR130078_2.fastq\n",
      "Approx 85% complete for DRR130078_2.fastq\n",
      "Approx 90% complete for DRR130078_2.fastq\n",
      "Approx 95% complete for DRR130078_2.fastq\n",
      "null\n",
      "null\n",
      "Analysis complete for DRR130093_1.fastq\n",
      "Analysis complete for DRR130093_2.fastq\n",
      "Started analysis of DRR130093_1.fastq\n",
      "Approx 5% complete for DRR130093_1.fastq\n",
      "Approx 10% complete for DRR130093_1.fastq\n",
      "Approx 15% complete for DRR130093_1.fastq\n",
      "Approx 20% complete for DRR130093_1.fastq\n",
      "Approx 25% complete for DRR130093_1.fastq\n",
      "Approx 30% complete for DRR130093_1.fastq\n",
      "Approx 35% complete for DRR130093_1.fastq\n",
      "Approx 40% complete for DRR130093_1.fastq\n",
      "Approx 45% complete for DRR130093_1.fastq\n",
      "Approx 50% complete for DRR130093_1.fastq\n",
      "Approx 55% complete for DRR130093_1.fastq\n",
      "Approx 60% complete for DRR130093_1.fastq\n",
      "Approx 65% complete for DRR130093_1.fastq\n",
      "Approx 70% complete for DRR130093_1.fastq\n",
      "Approx 75% complete for DRR130093_1.fastq\n",
      "Approx 80% complete for DRR130093_1.fastq\n",
      "Approx 85% complete for DRR130093_1.fastq\n",
      "Approx 90% complete for DRR130093_1.fastq\n",
      "Approx 95% complete for DRR130093_1.fastq\n",
      "Started analysis of DRR130093_2.fastq\n",
      "Approx 5% complete for DRR130093_2.fastq\n",
      "Approx 10% complete for DRR130093_2.fastq\n",
      "Approx 15% complete for DRR130093_2.fastq\n",
      "Approx 20% complete for DRR130093_2.fastq\n",
      "Approx 25% complete for DRR130093_2.fastq\n",
      "Approx 30% complete for DRR130093_2.fastq\n",
      "Approx 35% complete for DRR130093_2.fastq\n",
      "Approx 40% complete for DRR130093_2.fastq\n",
      "Approx 45% complete for DRR130093_2.fastq\n",
      "Approx 50% complete for DRR130093_2.fastq\n",
      "Approx 55% complete for DRR130093_2.fastq\n",
      "Approx 60% complete for DRR130093_2.fastq\n",
      "Approx 65% complete for DRR130093_2.fastq\n",
      "Approx 70% complete for DRR130093_2.fastq\n",
      "Approx 75% complete for DRR130093_2.fastq\n",
      "Approx 80% complete for DRR130093_2.fastq\n",
      "Approx 85% complete for DRR130093_2.fastq\n",
      "Approx 90% complete for DRR130093_2.fastq\n",
      "Approx 95% complete for DRR130093_2.fastq\n"
     ]
    }
   ],
   "source": [
    "quality_control ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "metadata": {},
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "trimming_bbduk () {\n",
    "  echo 'Read trimming with BBDuk'\n",
    "  local sample_file=\"$1\"\n",
    "  sed -i 's/\\r$//' \"$sample_file\"  \n",
    "\n",
    "  if [ ! -f \"$sample_file\" ]; then\n",
    "      echo \"Sample '$sample_file' file not found.\"\n",
    "      return 1\n",
    "  fi\n",
    "\n",
    "  cat \"$sample_file\" | xargs -I {} -P \"$(nproc)\" bash -c '\n",
    "    sample={};\n",
    "    echo \"Sample: $sample\";\n",
    "    if [ -d \"$HOME/Projeto_Mtb/Sample_reads/$sample\" ]; then\n",
    "      if [ -f \"$HOME/Projeto_Mtb/Sample_reads/$sample/${sample}_1.fastq\" ] && [ -f \"$HOME/Projeto_Mtb/Sample_reads/$sample/${sample}_2.fastq\" ]; then\n",
    "        bbduk.sh \\\n",
    "          in1=\"$HOME/Projeto_Mtb/Sample_reads/$sample/${sample}_1.fastq\" in2=\"$HOME/Projeto_Mtb/Sample_reads/$sample/${sample}_2.fastq\" \\\n",
    "          out=\"$HOME/Projeto_Mtb/Sample_reads/$sample/${sample}_R1_bbduk.fastq\" out2=\"$HOME/Projeto_Mtb/Sample_reads/$sample/${sample}_R2_bbduk.fastq\" \\\n",
    "          overwrite=t \\\n",
    "          ref=~/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta \\\n",
    "          ftm=5 \\\n",
    "          ktrim=r k=19 mink=8 editdistance=1 editdistance2=1 \\\n",
    "          trimpairsevenly=f removeifeitherbad=t \\\n",
    "          qtrim=r trimq=20 \\\n",
    "          trimpolygright=10 \\\n",
    "          minavgquality=20 \\\n",
    "          minlength=20 ottm=t \\\n",
    "          rename=t ziplevel=1 showspeed=t ;\n",
    "        rm \"$HOME/Projeto_Mtb/Sample_reads/$sample/${sample}\"_[1-2].fastq ; \n",
    "        echo \"Trimming for $sample completed\"\n",
    "      else\n",
    "        echo \"FASTQ not found for: $sample.\";\n",
    "      fi\n",
    "    else\n",
    "      echo \"Directory for $sample not found.\";\n",
    "    fi\n",
    "  '\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "metadata": {},
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read trimming with BBDuk\n",
      "Sample: DRR130077\n",
      "Sample: DRR130078\n",
      "Sample: DRR034421\n",
      "Sample: DRR130093\n",
      "/usr/share/bbmap/calcmem.sh: line 207: dpkg-architecture: command not found\n",
      "/usr/share/bbmap/calcmem.sh: line 207: dpkg-architecture: command not found\n",
      "/usr/share/bbmap/calcmem.sh: line 207: dpkg-architecture: command not found\n",
      "/usr/share/bbmap/calcmem.sh: line 207: dpkg-architecture: command not found\n",
      "java -ea -Xmx283m -Xms283m -cp /usr/share/java/bbmap.jar jgi.BBDuk in1=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_1.fastq in2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_2.fastq out=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_R1_bbduk.fastq out2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_R2_bbduk.fastq overwrite=t ref=/home/ritanobrega00/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta ftm=5 ktrim=r k=19 mink=8 editdistance=1 editdistance2=1 trimpairsevenly=f removeifeitherbad=t qtrim=r trimq=20 trimpolygright=10 minavgquality=20 minlength=20 ottm=t rename=t ziplevel=1 showspeed=t\n",
      "java -ea -Xmx283m -Xms283m -cp /usr/share/java/bbmap.jar jgi.BBDuk in1=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_1.fastq in2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_2.fastq out=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_R1_bbduk.fastq out2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_R2_bbduk.fastq overwrite=t ref=/home/ritanobrega00/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta ftm=5 ktrim=r k=19 mink=8 editdistance=1 editdistance2=1 trimpairsevenly=f removeifeitherbad=t qtrim=r trimq=20 trimpolygright=10 minavgquality=20 minlength=20 ottm=t rename=t ziplevel=1 showspeed=t\n",
      "java -ea -Xmx285m -Xms285m -cp /usr/share/java/bbmap.jar jgi.BBDuk in1=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_1.fastq in2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_2.fastq out=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_R1_bbduk.fastq out2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_R2_bbduk.fastq overwrite=t ref=/home/ritanobrega00/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta ftm=5 ktrim=r k=19 mink=8 editdistance=1 editdistance2=1 trimpairsevenly=f removeifeitherbad=t qtrim=r trimq=20 trimpolygright=10 minavgquality=20 minlength=20 ottm=t rename=t ziplevel=1 showspeed=t\n",
      "java -ea -Xmx285m -Xms285m -cp /usr/share/java/bbmap.jar jgi.BBDuk in1=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_1.fastq in2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_2.fastq out=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_R1_bbduk.fastq out2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_R2_bbduk.fastq overwrite=t ref=/home/ritanobrega00/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta ftm=5 ktrim=r k=19 mink=8 editdistance=1 editdistance2=1 trimpairsevenly=f removeifeitherbad=t qtrim=r trimq=20 trimpolygright=10 minavgquality=20 minlength=20 ottm=t rename=t ziplevel=1 showspeed=t\n",
      "Executing jgi.BBDuk [in1=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_1.fastq, in2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_2.fastq, out=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_R1_bbduk.fastq, out2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_R2_bbduk.fastq, overwrite=t, ref=/home/ritanobrega00/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta, ftm=5, ktrim=r, k=19, mink=8, editdistance=1, editdistance2=1, trimpairsevenly=f, removeifeitherbad=t, qtrim=r, trimq=20, trimpolygright=10, minavgquality=20, minlength=20, ottm=t, rename=t, ziplevel=1, showspeed=t]Executing jgi.BBDuk [in1=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_1.fastq, in2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_2.fastq, out=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_R1_bbduk.fastq, out2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_R2_bbduk.fastq, overwrite=t, ref=/home/ritanobrega00/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta, ftm=5, ktrim=r, k=19, mink=8, editdistance=1, editdistance2=1, trimpairsevenly=f, removeifeitherbad=t, qtrim=r, trimq=20, trimpolygright=10, minavgquality=20, minlength=20, ottm=t, rename=t, ziplevel=1, showspeed=t]\n",
      "\n",
      "Version 38.95Version 38.95Executing jgi.BBDuk [in1=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_1.fastq, in2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_2.fastq, out=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_R1_bbduk.fastq, out2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_R2_bbduk.fastq, overwrite=t, ref=/home/ritanobrega00/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta, ftm=5, ktrim=r, k=19, mink=8, editdistance=1, editdistance2=1, trimpairsevenly=f, removeifeitherbad=t, qtrim=r, trimq=20, trimpolygright=10, minavgquality=20, minlength=20, ottm=t, rename=t, ziplevel=1, showspeed=t]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Version 38.95\n",
      "\n",
      "Executing jgi.BBDuk [in1=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_1.fastq, in2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_2.fastq, out=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_R1_bbduk.fastq, out2=/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_R2_bbduk.fastq, overwrite=t, ref=/home/ritanobrega00/Projeto_Mtb/NGS_helper_files/adapters_combined_256_unique.fasta, ftm=5, ktrim=r, k=19, mink=8, editdistance=1, editdistance2=1, trimpairsevenly=f, removeifeitherbad=t, qtrim=r, trimq=20, trimpolygright=10, minavgquality=20, minlength=20, ottm=t, rename=t, ziplevel=1, showspeed=t]\n",
      "Version 38.95\n",
      "\n",
      "maskMiddle was disabled because useShortKmers=true\n",
      "maskMiddle was disabled because useShortKmers=true\n",
      "maskMiddle was disabled because useShortKmers=true\n",
      "maskMiddle was disabled because useShortKmers=true\n",
      "0.060 seconds.\n",
      "0.067 seconds.Initial:\n",
      "\n",
      "Memory: max=297m, total=297m, free=279m, used=18m\n",
      "Initial:\n",
      "\n",
      "Memory: max=297m, total=297m, free=279m, used=18m\n",
      "\n",
      "0.057 seconds.0.056 seconds.\n",
      "\n",
      "Initial:Initial:\n",
      "\n",
      "Memory: max=299m, total=299m, free=281m, used=18mMemory: max=299m, total=299m, free=281m, used=18m\n",
      "\n",
      "\n",
      "\n",
      "Added 380322 kmers; time: \t0.984 seconds.\n",
      "Memory: max=299m, total=299m, free=278m, used=21m\n",
      "\n",
      "Input is being processed as paired\n",
      "Added 380322 kmers; time: \t0.990 seconds.\n",
      "Memory: max=299m, total=299m, free=278m, used=21m\n",
      "\n",
      "Input is being processed as paired\n",
      "Added 380322 kmers; time: \t0.997 seconds.\n",
      "Memory: max=297m, total=297m, free=276m, used=21m\n",
      "\n",
      "Input is being processed as paired\n",
      "Added 380322 kmers; time: \t1.023 seconds.\n",
      "Memory: max=297m, total=297m, free=276m, used=21m\n",
      "\n",
      "Started output streams:\t0.052 seconds.\n",
      "Started output streams:\t0.073 seconds.\n",
      "Input is being processed as paired\n",
      "Started output streams:\t0.082 seconds.\n",
      "Started output streams:\t0.199 seconds.\n",
      "Processing time:   \t\t67.546 seconds.\n",
      "\n",
      "Input:                  \t1212518 reads \t\t304753833 bases.\n",
      "QTrimmed:               \t926237 reads (76.39%) \t100246514 bases (32.89%)\n",
      "Polymer-trimmed:        \t22 reads (0.00%) \t240 bases (0.00%)\n",
      "FTrimmed:               \t944697 reads (77.91%) \t1681978 bases (0.55%)\n",
      "KTrimmed:               \t182800 reads (15.08%) \t4981126 bases (1.63%)\n",
      "Low quality discards:   \t77106 reads (6.36%) \t8317898 bases (2.73%)\n",
      "Total Removed:          \t95148 reads (7.85%) \t115227756 bases (37.81%)\n",
      "Result:                 \t1117370 reads (92.15%) \t189526077 bases (62.19%)\n",
      "\n",
      "Time:                         \t68.790 seconds.\n",
      "Reads Processed:       1212k \t17.63k reads/sec\n",
      "Bases Processed:        304m \t4.43m bases/sec\n",
      "Trimming for DRR034421 completed\n",
      "Processing time:   \t\t72.361 seconds.\n",
      "\n",
      "Input:                  \t1147940 reads \t\t335422522 bases.\n",
      "QTrimmed:               \t1053575 reads (91.78%) \t93415295 bases (27.85%)\n",
      "Polymer-trimmed:        \t146 reads (0.01%) \t1858 bases (0.00%)\n",
      "FTrimmed:               \t287828 reads (25.07%) \t1000277 bases (0.30%)\n",
      "KTrimmed:               \t163951 reads (14.28%) \t13551186 bases (4.04%)\n",
      "Low quality discards:   \t32862 reads (2.86%) \t3689640 bases (1.10%)\n",
      "Total Removed:          \t79388 reads (6.92%) \t111658256 bases (33.29%)\n",
      "Result:                 \t1068552 reads (93.08%) \t223764266 bases (66.71%)\n",
      "\n",
      "Time:                         \t73.447 seconds.\n",
      "Reads Processed:       1147k \t15.63k reads/sec\n",
      "Bases Processed:        335m \t4.57m bases/sec\n",
      "Trimming for DRR130077 completed\n",
      "Processing time:   \t\t85.569 seconds.\n",
      "\n",
      "Input:                  \t1727074 reads \t\t497995063 bases.\n",
      "QTrimmed:               \t1421317 reads (82.30%) \t68619895 bases (13.78%)\n",
      "Polymer-trimmed:        \t117 reads (0.01%) \t1384 bases (0.00%)\n",
      "FTrimmed:               \t450461 reads (26.08%) \t1448788 bases (0.29%)\n",
      "KTrimmed:               \t220847 reads (12.79%) \t7707993 bases (1.55%)\n",
      "Low quality discards:   \t30334 reads (1.76%) \t3792020 bases (0.76%)\n",
      "Total Removed:          \t56582 reads (3.28%) \t81570080 bases (16.38%)\n",
      "Result:                 \t1670492 reads (96.72%) \t416424983 bases (83.62%)\n",
      "\n",
      "Time:                         \t86.650 seconds.\n",
      "Reads Processed:       1727k \t19.93k reads/sec\n",
      "Bases Processed:        497m \t5.75m bases/sec\n",
      "Trimming for DRR130078 completed\n",
      "Processing time:   \t\t98.712 seconds.\n",
      "\n",
      "Input:                  \t2736154 reads \t\t795534093 bases.\n",
      "QTrimmed:               \t2335328 reads (85.35%) \t135818388 bases (17.07%)\n",
      "Polymer-trimmed:        \t271 reads (0.01%) \t3335 bases (0.00%)\n",
      "FTrimmed:               \t691445 reads (25.27%) \t2133908 bases (0.27%)\n",
      "KTrimmed:               \t325677 reads (11.90%) \t4310091 bases (0.54%)\n",
      "Low quality discards:   \t74454 reads (2.72%) \t9408039 bases (1.18%)\n",
      "Total Removed:          \t89050 reads (3.25%) \t151673761 bases (19.07%)\n",
      "Result:                 \t2647104 reads (96.75%) \t643860332 bases (80.93%)\n",
      "\n",
      "Time:                         \t99.767 seconds.\n",
      "Reads Processed:       2736k \t27.43k reads/sec\n",
      "Bases Processed:        795m \t7.97m bases/sec\n",
      "Trimming for DRR130093 completed\n"
     ]
    }
   ],
   "source": [
    "trimming_bbduk ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **3. Mapping**\n",
    "\n",
    "After the pre-processement to ensure we will only work with high-quality reads, the trimmed reads will be mapped against the *Mtb*'s ancestral reference genome.\n",
    "\n",
    "The function `map_bwa` uses the `bwa mem` [3] and `samtools` to perform the mapping and consequent creation of the BAM file per sample, which will be indexed, sorted by read name, and compressed, to facilitate efficient access and retrieval of read alignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "map_bwa() {\n",
    "  echo \"Mapping genomes to MTB_anc with bwa mem + sorting BAM by read name\"\n",
    "  \n",
    "  local sample_file=\"$1\"\n",
    "  local genome_reference=\"$HOME/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta\"\n",
    "  local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "\n",
    "  if [ ! -f \"$sample_file\" ]; then\n",
    "      echo \"Sample '$sample_file' file not found.\"\n",
    "      return 1\n",
    "  fi\n",
    "\n",
    "  while IFS= read -r sample; do\n",
    "    echo -e \"\\nMapping Sample: $sample\"\n",
    "    \n",
    "    if [ ! -f \"$sra_diretoria/$sample/${sample}_R1_bbduk.fastq\" ] || [ ! -f \"$sra_diretoria/$sample/${sample}_R2_bbduk.fastq\" ]; then\n",
    "      echo \"Input for '$sample_file' not found.\"\n",
    "      continue\n",
    "    fi\n",
    "\n",
    "    bwa_threads=$(nproc)\n",
    "    bwa mem -t \"$bwa_threads\" \"$genome_reference\" \"$sra_diretoria/$sample/${sample}_R1_bbduk.fastq\" \"$sra_diretoria/$sample/${sample}_R2_bbduk.fastq\" \\\n",
    "    | samtools sort -n -l 1 -@ 1 -o \"$sra_diretoria/$sample/$sample.bam\"\n",
    "\n",
    "    samtools view -@ 1 -b \"$sra_diretoria/$sample/$sample.bam\" | samtools sort -@ 1 -o \"$sra_diretoria/$sample/${sample}_sorted.bam\"\n",
    "    samtools index -@ 1 \"$sra_diretoria/$sample/${sample}_sorted.bam\"\n",
    "    \n",
    "    rm \"$sra_diretoria/$sample/${sample}_R1_bbduk.fastq\" \"$sra_diretoria/$sample/${sample}_R2_bbduk.fastq\"\n",
    "\n",
    "    echo \"Mapping for $sample completed.\"\n",
    "  done < \"$sample_file\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping genomes to MTB_anc with bwa mem + sorting BAM by read name\n",
      "\n",
      "Mapping Sample: DRR034421\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 461778 sequences (80000283 bp)...\n",
      "[M::process] read 476458 sequences (80000462 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (843, 221806, 87, 561)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (117, 200, 357)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 837)\n",
      "[M::mem_pestat] mean and std.dev: (242.04, 171.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1077)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (178, 266, 389)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 811)\n",
      "[M::mem_pestat] mean and std.dev: (294.03, 155.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1022)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (185, 641, 1765)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4925)\n",
      "[M::mem_pestat] mean and std.dev: (1091.02, 1104.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6505)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (125, 228, 391)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 923)\n",
      "[M::mem_pestat] mean and std.dev: (263.61, 188.51)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1189)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 461778 reads in 107.856 CPU sec, 15.043 real sec\n",
      "[M::process] read 179134 sequences (29525332 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1035, 227996, 89, 573)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (142, 234, 389)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 883)\n",
      "[M::mem_pestat] mean and std.dev: (274.60, 182.07)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1130)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (189, 277, 401)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 825)\n",
      "[M::mem_pestat] mean and std.dev: (305.47, 157.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1037)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (310, 805, 2148)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5824)\n",
      "[M::mem_pestat] mean and std.dev: (1225.48, 1110.98)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7662)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (147, 234, 448)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1050)\n",
      "[M::mem_pestat] mean and std.dev: (282.26, 197.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1351)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 476458 reads in 103.725 CPU sec, 13.723 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (381, 85610, 49, 232)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (147, 238, 406)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 924)\n",
      "[M::mem_pestat] mean and std.dev: (283.48, 185.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1183)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (204, 294, 423)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 861)\n",
      "[M::mem_pestat] mean and std.dev: (324.75, 162.26)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1080)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 773, 1808)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4962)\n",
      "[M::mem_pestat] mean and std.dev: (1089.82, 1061.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6539)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (166, 251, 405)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 883)\n",
      "[M::mem_pestat] mean and std.dev: (278.08, 166.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1122)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 179134 reads in 42.247 CPU sec, 5.772 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -t 8 /home/ritanobrega00/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_R1_bbduk.fastq /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_R2_bbduk.fastq\n",
      "[main] Real time: 37.049 sec; CPU: 255.750 sec\n",
      "Mapping for DRR034421 completed.\n",
      "\n",
      "Mapping Sample: DRR130077\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 390894 sequences (80000342 bp)...\n",
      "[M::process] read 381678 sequences (80000403 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (581, 185285, 90, 520)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 303, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 658)\n",
      "[M::mem_pestat] mean and std.dev: (303.31, 99.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 797)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (383, 426, 477)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (195, 665)\n",
      "[M::mem_pestat] mean and std.dev: (428.43, 75.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (101, 759)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (690, 1837, 2430)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5910)\n",
      "[M::mem_pestat] mean and std.dev: (1723.08, 1260.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7650)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (200, 272, 360)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 680)\n",
      "[M::mem_pestat] mean and std.dev: (272.41, 110.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 840)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 390894 reads in 185.388 CPU sec, 31.614 real sec\n",
      "[M::process] read 295980 sequences (63763521 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (578, 181153, 67, 475)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (227, 304, 385)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 701)\n",
      "[M::mem_pestat] mean and std.dev: (302.69, 113.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 859)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (384, 428, 480)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (192, 672)\n",
      "[M::mem_pestat] mean and std.dev: (430.06, 76.26)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (96, 768)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1250, 1968, 2524)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5072)\n",
      "[M::mem_pestat] mean and std.dev: (1763.23, 1055.72)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6346)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (201, 270, 351)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 651)\n",
      "[M::mem_pestat] mean and std.dev: (273.69, 108.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 801)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 381678 reads in 163.378 CPU sec, 22.446 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (436, 140655, 50, 336)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 303, 366)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 628)\n",
      "[M::mem_pestat] mean and std.dev: (296.24, 104.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 759)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (385, 429, 482)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (191, 676)\n",
      "[M::mem_pestat] mean and std.dev: (431.08, 76.48)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 773)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (810, 1617, 2370)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5490)\n",
      "[M::mem_pestat] mean and std.dev: (1627.44, 1177.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7050)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (200, 272, 377)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 731)\n",
      "[M::mem_pestat] mean and std.dev: (278.66, 121.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 908)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 295980 reads in 162.043 CPU sec, 24.860 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -t 8 /home/ritanobrega00/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_R1_bbduk.fastq /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_R2_bbduk.fastq\n",
      "[main] Real time: 113.816 sec; CPU: 539.470 sec\n",
      "Mapping for DRR130077 completed.\n",
      "\n",
      "Mapping Sample: DRR130078\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 324350 sequences (80000298 bp)...\n",
      "[M::process] read 323674 sequences (80000401 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (279, 156271, 31, 185)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (193, 236, 293)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 493)\n",
      "[M::mem_pestat] mean and std.dev: (239.90, 74.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 593)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (347, 377, 411)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (219, 539)\n",
      "[M::mem_pestat] mean and std.dev: (378.48, 48.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (155, 603)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1294, 1559, 2434)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4714)\n",
      "[M::mem_pestat] mean and std.dev: (1555.21, 919.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5854)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (150, 203, 245)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 435)\n",
      "[M::mem_pestat] mean and std.dev: (192.25, 69.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 530)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 324350 reads in 225.708 CPU sec, 32.863 real sec\n",
      "[M::process] read 320878 sequences (80000165 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (337, 156061, 34, 227)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (201, 251, 306)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 516)\n",
      "[M::mem_pestat] mean and std.dev: (248.87, 72.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 621)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (347, 377, 411)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (219, 539)\n",
      "[M::mem_pestat] mean and std.dev: (378.32, 48.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (155, 603)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (493, 1396, 2222)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5680)\n",
      "[M::mem_pestat] mean and std.dev: (1355.18, 951.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7409)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (152, 203, 273)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 515)\n",
      "[M::mem_pestat] mean and std.dev: (209.72, 89.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 636)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 323674 reads in 230.522 CPU sec, 32.329 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (298, 154697, 41, 222)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (203, 250, 304)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 506)\n",
      "[M::mem_pestat] mean and std.dev: (249.56, 76.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 607)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (347, 377, 411)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (219, 539)\n",
      "[M::mem_pestat] mean and std.dev: (378.71, 48.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (155, 603)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (169, 1289, 2370)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6772)\n",
      "[M::mem_pestat] mean and std.dev: (1471.00, 1537.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8973)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (157, 203, 263)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 475)\n",
      "[M::mem_pestat] mean and std.dev: (203.73, 70.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 581)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 320878 reads in 246.114 CPU sec, 47.606 real sec\n",
      "[M::process] read 318832 sequences (80000083 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (284, 153717, 47, 208)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (196, 240, 292)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (4, 484)\n",
      "[M::mem_pestat] mean and std.dev: (238.30, 73.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 580)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (348, 377, 412)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (220, 540)\n",
      "[M::mem_pestat] mean and std.dev: (379.04, 48.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (156, 604)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (415, 914, 1664)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4162)\n",
      "[M::mem_pestat] mean and std.dev: (1138.83, 938.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5411)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (162, 215, 280)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 516)\n",
      "[M::mem_pestat] mean and std.dev: (215.57, 85.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 634)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 318832 reads in 225.924 CPU sec, 32.439 real sec\n",
      "[M::process] read 318064 sequences (80000368 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (293, 153329, 46, 202)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (204, 246, 299)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (14, 489)\n",
      "[M::mem_pestat] mean and std.dev: (243.66, 72.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 584)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (348, 377, 412)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (220, 540)\n",
      "[M::mem_pestat] mean and std.dev: (378.97, 48.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (156, 604)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (425, 1396, 2153)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5609)\n",
      "[M::mem_pestat] mean and std.dev: (1253.67, 974.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7337)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (162, 202, 261)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 459)\n",
      "[M::mem_pestat] mean and std.dev: (205.38, 74.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 558)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 318064 reads in 205.055 CPU sec, 32.199 real sec\n",
      "[M::process] read 64694 sequences (16423668 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (61, 31190, 12, 32)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (180, 236, 294)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 522)\n",
      "[M::mem_pestat] mean and std.dev: (239.67, 69.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 636)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (349, 378, 413)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (221, 541)\n",
      "[M::mem_pestat] mean and std.dev: (379.35, 48.71)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (157, 605)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1635, 2215, 3203)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6339)\n",
      "[M::mem_pestat] mean and std.dev: (2064.64, 1259.72)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7907)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (168, 223, 315)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 609)\n",
      "[M::mem_pestat] mean and std.dev: (207.77, 91.28)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 756)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 64694 reads in 26.025 CPU sec, 3.878 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -t 8 /home/ritanobrega00/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_R1_bbduk.fastq /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_R2_bbduk.fastq\n",
      "[main] Real time: 256.137 sec; CPU: 1184.592 sec\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "Mapping for DRR130078 completed.\n",
      "\n",
      "Mapping Sample: DRR130093\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 352688 sequences (80000082 bp)...\n",
      "[M::process] read 346144 sequences (80000239 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (374, 167792, 38, 281)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (216, 270, 329)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 555)\n",
      "[M::mem_pestat] mean and std.dev: (265.07, 76.33)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 668)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (371, 396, 425)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (263, 533)\n",
      "[M::mem_pestat] mean and std.dev: (400.13, 41.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (209, 587)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (913, 1636, 2492)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5650)\n",
      "[M::mem_pestat] mean and std.dev: (1864.24, 1286.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7229)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (179, 223, 278)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 476)\n",
      "[M::mem_pestat] mean and std.dev: (220.40, 70.44)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 575)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 352688 reads in 120.288 CPU sec, 17.162 real sec\n",
      "[M::process] read 325702 sequences (80000300 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (360, 164822, 38, 254)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (216, 273, 322)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (4, 534)\n",
      "[M::mem_pestat] mean and std.dev: (267.47, 75.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 640)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (371, 396, 425)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (263, 533)\n",
      "[M::mem_pestat] mean and std.dev: (400.23, 40.95)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (209, 587)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1509, 2207, 2499)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4479)\n",
      "[M::mem_pestat] mean and std.dev: (1908.40, 911.77)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5555)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (161, 217, 263)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 467)\n",
      "[M::mem_pestat] mean and std.dev: (211.06, 72.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 569)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 346144 reads in 113.530 CPU sec, 15.014 real sec\n",
      "[M::process] read 326172 sequences (80000341 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (268, 156529, 36, 215)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (188, 229, 278)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (8, 458)\n",
      "[M::mem_pestat] mean and std.dev: (234.46, 65.61)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 548)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (331, 358, 388)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (217, 502)\n",
      "[M::mem_pestat] mean and std.dev: (360.21, 42.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (160, 559)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (349, 2196, 2523)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6871)\n",
      "[M::mem_pestat] mean and std.dev: (1743.43, 1444.48)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 9045)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (149, 194, 265)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
      "[M::mem_pestat] mean and std.dev: (198.95, 77.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 325702 reads in 100.648 CPU sec, 13.198 real sec\n",
      "[M::process] read 318534 sequences (80000384 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (283, 156843, 27, 193)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (180, 235, 293)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 519)\n",
      "[M::mem_pestat] mean and std.dev: (234.65, 75.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 632)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (330, 355, 383)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (224, 489)\n",
      "[M::mem_pestat] mean and std.dev: (357.11, 39.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (171, 542)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (1348, 2125, 2430)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4594)\n",
      "[M::mem_pestat] mean and std.dev: (1688.23, 821.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5676)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (145, 189, 258)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 484)\n",
      "[M::mem_pestat] mean and std.dev: (200.62, 78.81)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 597)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 326172 reads in 107.497 CPU sec, 14.106 real sec\n",
      "[M::process] read 323180 sequences (80000173 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (229, 153426, 29, 182)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (188, 221, 286)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 482)\n",
      "[M::mem_pestat] mean and std.dev: (227.58, 61.33)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 580)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (330, 355, 384)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (222, 492)\n",
      "[M::mem_pestat] mean and std.dev: (357.28, 39.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (168, 546)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (886, 2182, 2904)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6940)\n",
      "[M::mem_pestat] mean and std.dev: (1907.83, 1318.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8958)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (151, 197, 253)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 457)\n",
      "[M::mem_pestat] mean and std.dev: (194.23, 66.52)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 559)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 318534 reads in 116.585 CPU sec, 15.850 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (253, 155523, 17, 199)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (191, 236, 287)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 479)\n",
      "[M::mem_pestat] mean and std.dev: (235.12, 70.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 575)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (330, 355, 384)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (222, 492)\n",
      "[M::mem_pestat] mean and std.dev: (357.37, 39.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (168, 546)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (738, 1389, 2499)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6021)\n",
      "[M::mem_pestat] mean and std.dev: (1574.82, 928.23)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7782)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (148, 183, 226)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 382)\n",
      "[M::mem_pestat] mean and std.dev: (186.29, 62.33)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 460)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 323180 reads in 118.121 CPU sec, 17.413 real sec\n",
      "[M::process] read 323400 sequences (80000298 bp)...\n",
      "[M::process] read 316230 sequences (80000046 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (289, 155600, 22, 201)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (190, 233, 291)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 493)\n",
      "[M::mem_pestat] mean and std.dev: (235.07, 66.56)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 594)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (330, 355, 384)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (222, 492)\n",
      "[M::mem_pestat] mean and std.dev: (357.20, 39.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (168, 546)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (921, 1325, 2253)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4917)\n",
      "[M::mem_pestat] mean and std.dev: (1371.90, 808.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6249)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (141, 188, 239)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 435)\n",
      "[M::mem_pestat] mean and std.dev: (188.30, 68.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 533)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 323400 reads in 98.707 CPU sec, 12.775 real sec\n",
      "[M::process] read 15054 sequences (3858469 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (246, 152364, 23, 184)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (188, 231, 284)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 476)\n",
      "[M::mem_pestat] mean and std.dev: (229.85, 70.52)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 572)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (331, 356, 384)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (225, 490)\n",
      "[M::mem_pestat] mean and std.dev: (357.69, 39.02)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (172, 543)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (857, 1313, 2238)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5000)\n",
      "[M::mem_pestat] mean and std.dev: (1326.96, 745.54)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6381)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (139, 185, 242)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 448)\n",
      "[M::mem_pestat] mean and std.dev: (187.61, 76.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 551)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 316230 reads in 108.064 CPU sec, 14.139 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 7257, 1, 6)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (212, 256, 297)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (42, 467)\n",
      "[M::mem_pestat] mean and std.dev: (248.64, 50.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 552)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (331, 356, 384)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (225, 490)\n",
      "[M::mem_pestat] mean and std.dev: (357.00, 39.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (172, 543)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_process_seqs] Processed 15054 reads in 5.126 CPU sec, 0.724 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -t 8 /home/ritanobrega00/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_R1_bbduk.fastq /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_R2_bbduk.fastq\n",
      "[main] Real time: 140.859 sec; CPU: 892.638 sec\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "Mapping for DRR130093 completed.\n"
     ]
    }
   ],
   "source": [
    "map_bwa ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **4. Post-Processing**\n",
    "\n",
    "After mapping, mapped reads will be post-processed to ensure the reliability of the subsequent analysis (variant calling).\n",
    "\n",
    "First, duplicates will be marked using `samtools markup` in the function `mark_duplicates`, and the BAM files will be indexed.\n",
    "\n",
    "Then, for each of these BAM files, coverage will be calculated using `mosdepth` [4] to provide a measure of the depth of sequencing across the genome. In the function `calculate_bam_coverage`, only properly paired and mapped reads will be included in the coverage calculation. Moreover, only reads with a mapping quality of at least 20 will be included (the median will be used instead of the mean) and, for each sample, 4 files will be created: summary; golbal.dist; per-base.bed (zip and csi file).\n",
    "\n",
    "After obtaining BAM coverage, the function `process_coverage_metrics` will extract and store (in a file) the coverage metrics, as well as remove and store the regions with no coverage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# Duplicates marking \n",
    "mark_duplicates() {\n",
    "  echo \"Marking duplicates with samtools markdup and indexing BAM files\"\n",
    "  local sample_file=\"$1\"\n",
    "  local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "  local threads=$(nproc)\n",
    "\n",
    "  if [ ! -f \"$sample_file\" ]; then\n",
    "    echo \"Sample '$sample_file' file not found.\"\n",
    "    return 1\n",
    "  fi\n",
    "\n",
    "  while IFS= read -r sample || [[ -n $sample ]]; do\n",
    "    if [ ! -f \"$sra_diretoria/$sample/$sample.bam\" ]; then\n",
    "      echo \"No BAM file found for: $sample.\"\n",
    "      continue\n",
    "    fi\n",
    "\n",
    "    samtools fixmate -m -@ 2 \"$sra_diretoria/$sample/$sample.bam\" -u - \\\n",
    "    | samtools sort -u -@ 2 - \\\n",
    "    | samtools markdup --include-fails -S --mode s -@ 2 - -O bam,level=1 \"$sra_diretoria/$sample/${sample}_markdup.bam\" \\\n",
    "    && samtools index -@ 2 \"$sra_diretoria/$sample/${sample}_markdup.bam\" \\\n",
    "    && rm \"$sra_diretoria/$sample/$sample.bam\"\n",
    "\n",
    "  done < \"$sample_file\"\n",
    "\n",
    "  echo \"Duplicate marking and indexing finished\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Marking duplicates with samtools markdup and indexing BAM files\n",
      "[bam_sort_core] merging from 0 files and 2 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 2 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 2 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 2 in-memory blocks...\n",
      "Duplicate marking and indexing finished\n"
     ]
    }
   ],
   "source": [
    "mark_duplicates ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "calculate_bam_coverage() {\n",
    "  local sample_file=\"$1\"\n",
    "  local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "  local threads=$(nproc)\n",
    "  local mosdepth_bin=\"$CONDA_PREFIX/bin/mosdepth\"\n",
    "\n",
    "  echo \"Calculating bam coverage with mosdepth in parallel\"\n",
    "  while IFS= read -r sample || [[ -n $sample ]]; do\n",
    "    if [ -z \"$sample\" ]; then\n",
    "      continue \n",
    "    fi\n",
    "\n",
    "    if [ ! -d \"$sra_diretoria/$sample\" ]; then\n",
    "      echo \"Directory not found for: $sample\"\n",
    "      continue\n",
    "    fi\n",
    "\n",
    "    bam_file=\"$sra_diretoria/$sample/${sample}_markdup.bam\"\n",
    "    if [ ! -f \"$bam_file\" ]; then\n",
    "      echo \"No BAM found for: $sample\"\n",
    "      continue\n",
    "    fi\n",
    "\n",
    "    pushd \"$sra_diretoria/$sample\" > /dev/null || continue\n",
    "    echo \"Calculating coverage for ${sample}_markdup.bam\"\n",
    "    echo \"$bam_file\" | parallel -j0 --colsep=\"\\t\" \"${mosdepth_bin}\" --flag 3844 --mapq 20 --use-median --threads \"$threads\" \"${sample}_markdup.bam\"\n",
    "\n",
    "    echo \"BAM coverage calculated for $sample\"\n",
    "\n",
    "    popd > /dev/null || continue\n",
    "  done < \"$sample_file\"\n",
    "\n",
    "  echo \"Coverage calculation completed\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating bam coverage with mosdepth in parallel\n",
      "Calculating coverage for DRR034421_markdup.bam\n",
      "Academic tradition requires you to cite works you base your article on.\n",
      "When using programs that use GNU Parallel to process data for publication\n",
      "please cite:\n",
      "\n",
      "  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,\n",
      "  ;login: The USENIX Magazine, February 2011:42-47.\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "To silence the citation notice: run 'parallel --citation'.\n",
      "\n",
      "BAM coverage calculated for DRR034421\n",
      "Calculating coverage for DRR130077_markdup.bam\n",
      "Academic tradition requires you to cite works you base your article on.\n",
      "When using programs that use GNU Parallel to process data for publication\n",
      "please cite:\n",
      "\n",
      "  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,\n",
      "  ;login: The USENIX Magazine, February 2011:42-47.\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "To silence the citation notice: run 'parallel --citation'.\n",
      "\n",
      "BAM coverage calculated for DRR130077\n",
      "Calculating coverage for DRR130078_markdup.bam\n",
      "Academic tradition requires you to cite works you base your article on.\n",
      "When using programs that use GNU Parallel to process data for publication\n",
      "please cite:\n",
      "\n",
      "  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,\n",
      "  ;login: The USENIX Magazine, February 2011:42-47.\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "To silence the citation notice: run 'parallel --citation'.\n",
      "\n",
      "BAM coverage calculated for DRR130078\n",
      "Calculating coverage for DRR130093_markdup.bam\n",
      "Academic tradition requires you to cite works you base your article on.\n",
      "When using programs that use GNU Parallel to process data for publication\n",
      "please cite:\n",
      "\n",
      "  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,\n",
      "  ;login: The USENIX Magazine, February 2011:42-47.\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "To silence the citation notice: run 'parallel --citation'.\n",
      "\n",
      "BAM coverage calculated for DRR130093\n",
      "Coverage calculation completed\n"
     ]
    }
   ],
   "source": [
    "calculate_bam_coverage ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "process_coverage_metrics() {\n",
    "    local sample_file=\"$1\"\n",
    "    local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "\n",
    "    while IFS= read -r sample || [[ -n $sample ]]; do\n",
    "        local summary_file=\"$sra_diretoria/$sample/${sample}_markdup.bam.mosdepth.summary.txt\"\n",
    "        local dist_file=\"$sra_diretoria/$sample/${sample}_markdup.bam.mosdepth.global.dist.txt\"\n",
    "        \n",
    "        if [ ! -f \"$summary_file\" ] || [ ! -f \"$dist_file\" ]; then\n",
    "            echo \"Coverage metric files not found for sample $sample\"\n",
    "            continue\n",
    "        fi\n",
    "\n",
    "        echo \"Extracting coverage metrics for $sample\"\n",
    "        \n",
    "        local median=$(awk 'NR>1 && NR%2==1 {print $4}' \"$summary_file\")\n",
    "        local cov10=$(grep -m1 '^10\\t' \"$dist_file\" | awk '{print $3}')\n",
    "        echo -e \"$sample\\t$median\\t$cov10\" > \"$sra_diretoria/$sample/${sample}_coveragecheck.tsv\"\n",
    "\n",
    "        echo \"Extracting genomic positions without mapped coverage for $sample\"\n",
    "        \n",
    "        grep -h -E '^[^[:space:]]+[[:space:]]+0$' \"$sra_diretoria/$sample/\"*per-base* > \"$sra_diretoria/$sample/${sample}_zerocoverageregions.tsv\"\n",
    "\n",
    "        rm -f \"$sra_diretoria/$sample/\"*mosdepth* \"$sra_diretoria/$sample/\"*per-base*\n",
    "    \n",
    "    done < \"$sample_file\" \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting coverage metrics for DRR034421\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting genomic positions without mapped coverage for DRR034421\n",
      "Extracting coverage metrics for DRR130077\n",
      "Extracting genomic positions without mapped coverage for DRR130077\n",
      "Extracting coverage metrics for DRR130078\n",
      "Extracting genomic positions without mapped coverage for DRR130078\n",
      "Extracting coverage metrics for DRR130093\n",
      "Extracting genomic positions without mapped coverage for DRR130093\n"
     ]
    }
   ],
   "source": [
    "process_coverage_metrics ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **5. Variant Calling**\n",
    "\n",
    "The processed BAM files will be used for the variant calling to ensure the use and call of realiable data.\n",
    "\n",
    "For variant identification, different strategies were implemented to compare and integrate the information of different callers in the final result.   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simple variant calling with `bcftools mpileup`\n",
    "\n",
    "Firstly, the most simple variant calling was performed using the `bcftools mpileup` and `bcftools call` commands with no extra options in order to obtain the most basic and simple VCF file [5]. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "variant_calling_simple() {\n",
    "    local sample_file=\"$1\"\n",
    "    local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "    local ref_file=\"$HOME/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta\"\n",
    "    \n",
    "    while IFS= read -r sample || [[ -n $sample ]]; do\n",
    "        echo \"Running variant calling for sample: $sample\" \n",
    "        bcftools mpileup -Ou -f \"$ref_file\" \"$sra_diretoria/$sample/${sample}_markdup.bam\" \\\n",
    "        | bcftools call --ploidy 1 -Ov -vc -o \"$sra_diretoria/$sample/${sample}_simple.vcf\"\n",
    "    done < \"$sample_file\" \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running variant calling for sample: DRR034421\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Running variant calling for sample: DRR130077\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Running variant calling for sample: DRR130078\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Running variant calling for sample: DRR130093\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n"
     ]
    }
   ],
   "source": [
    "variant_calling_simple ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variant Calling with Filters and Annotations using `bcftools mpileup, call, norm, annotate`\n",
    "\n",
    "The second strategy used consisted in the use of `bcftools mpileup` and `call`, as well as `norm` and `annotate` [5], to apply filters (e.g. use of reads with a quality of alignment and mapping superior to 20), normalize the representation of the variants called, and add information for each one regarding, for example, lineage, excluded regions and mapability. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "create_intervals() {\n",
    "    local threads=24\n",
    "    local output_file=\"$HOME/Projeto_Mtb/NGS_helper_files/intervals_${threads}threads.bed\"\n",
    "    \n",
    "    echo \"Creating equally-sized intervals file for the reference genome\"\n",
    "    bedtools makewindows -g ~/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta.fai -n \"$threads\" -i winnum \\\n",
    "    | awk '{print $1\"\\t\"$2+1\"\\t\"$3}' > \"$output_file\"\n",
    "    cat -n \"$output_file\"\n",
    "}\n",
    "\n",
    "run_variant_calling() {\n",
    "    local sample=\"$1\"\n",
    "    local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "    local output_dir=\"$sra_diretoria/$sample\"\n",
    "    local output_prefix=\"${sample}_bcftools_varsonly\"\n",
    "    local ref_file=\"$HOME/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta\"\n",
    "    local threads=$(nproc)\n",
    "    local intervals_file=\"$HOME/Projeto_Mtb/NGS_helper_files/intervals_24threads.bed\"\n",
    "\n",
    "    echo \"Running variant calling for sample: $sample\"\n",
    "    bcftools mpileup -f \"$ref_file\" \"$sra_diretoria/$sample/${sample}_markdup.bam\" \\\n",
    "    --count-orphans \\\n",
    "    --no-BAQ --min-MQ 20 --min-BQ 20 \\\n",
    "    --regions-file \"$intervals_file\" \\\n",
    "    --annotate AD,ADF,ADR,DP,SP,SCR,INFO/AD,INFO/ADF,INFO/ADR,INFO/SCR \\\n",
    "    --threads \"$threads\" --output-type u \\\n",
    "    | bcftools call --ploidy 1 \\\n",
    "    --keep-alts --keep-masked-ref \\\n",
    "    --multiallelic-caller \\\n",
    "    --variants-only \\\n",
    "    --threads \"$threads\" --output-type u \\\n",
    "    | bcftools norm --fasta-ref \"$ref_file\" \\\n",
    "    --multiallelics - --keep-sum AD \\\n",
    "    --threads \"$threads\" --output-type v \\\n",
    "    | bcftools annotate \\\n",
    "    --annotations ~/Projeto_Mtb/NGS_helper_files/excludedloci_RLC2021_annot.tab.gz \\\n",
    "    --header-lines ~/Projeto_Mtb/NGS_helper_files/excludedloci_RLC2021_annot.header \\\n",
    "    --columns CHROM,FROM,TO,RLC_tag \\\n",
    "    --threads \"$threads\" --output-type u \\\n",
    "    | bcftools annotate \\\n",
    "    --annotations ~/Projeto_Mtb/NGS_helper_files/blindspots_mappability_marin2021_annot.tab.gz \\\n",
    "    --header-lines ~/Projeto_Mtb/NGS_helper_files/blindspots_mappability_marin2021_annot.header \\\n",
    "    --columns CHROM,FROM,TO,Mappability \\\n",
    "    --threads \"$threads\" --output-type u \\\n",
    "    | bcftools annotate \\\n",
    "    --annotations ~/Projeto_Mtb/NGS_helper_files/lineagesnps_annot.tab.gz \\\n",
    "    --header-lines ~/Projeto_Mtb/NGS_helper_files/lineagesnps_annot.header \\\n",
    "    --columns CHROM,POS,REF,ALT,Lineage_tag \\\n",
    "    --threads \"$threads\" --output-type u \\\n",
    "    | bcftools annotate \\\n",
    "    --annotations ~/Projeto_Mtb/NGS_helper_files/iedbepitopes_annot.tab.gz \\\n",
    "    --header-lines ~/Projeto_Mtb/NGS_helper_files/iedbepitopes_annot.header \\\n",
    "    --columns CHROM,POS,REF,ALT,IEDB_tag \\\n",
    "    --merge-logic IEDB_tag:unique \\\n",
    "    --threads \"$threads\" --output-type v \\\n",
    "    | bcftools annotate --set-id +'%CHROM:%POS' \\\n",
    "    | bgzip > \"$output_dir/${output_prefix}_annotated.vcf.gz\"\n",
    "\n",
    "    if [ -s \"$output_dir/${output_prefix}_annotated.vcf.gz\" ]; then\n",
    "        echo \"Indexing the final VCF file\"\n",
    "        tabix -p vcf \"$output_dir/${output_prefix}_annotated.vcf.gz\"\n",
    "    else\n",
    "        echo \"Error: Failed to create the final VCF file for sample: $sample\"\n",
    "    fi\n",
    "}\n",
    "\n",
    "variant_calling() {\n",
    "    local sample_file=\"$1\"\n",
    "    local threads=$(nproc)\n",
    "    echo \"Creating equally-sized intervals file for the reference genome\"\n",
    "    create_intervals\n",
    "    \n",
    "    echo \"Variant calling with bcftools parallel jobs\"\n",
    "    while IFS= read -r sample || [[ -n $sample ]]; do\n",
    "        echo \"Running variant calling for sample: $sample\"\n",
    "        run_variant_calling \"$sample\"\n",
    "    done < \"$sample_file\"    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating equally-sized intervals file for the reference genome\n",
      "Creating equally-sized intervals file for the reference genome\n",
      "WARNING: Interval :0-0 is smaller than the number of windows requested. Skipping.\n",
      "     1\tMTB_anc\t1\t183813\n",
      "     2\tMTB_anc\t183814\t367626\n",
      "     3\tMTB_anc\t367627\t551439\n",
      "     4\tMTB_anc\t551440\t735252\n",
      "     5\tMTB_anc\t735253\t919065\n",
      "     6\tMTB_anc\t919066\t1102878\n",
      "     7\tMTB_anc\t1102879\t1286691\n",
      "     8\tMTB_anc\t1286692\t1470504\n",
      "     9\tMTB_anc\t1470505\t1654317\n",
      "    10\tMTB_anc\t1654318\t1838130\n",
      "    11\tMTB_anc\t1838131\t2021943\n",
      "    12\tMTB_anc\t2021944\t2205756\n",
      "    13\tMTB_anc\t2205757\t2389569\n",
      "    14\tMTB_anc\t2389570\t2573382\n",
      "    15\tMTB_anc\t2573383\t2757195\n",
      "    16\tMTB_anc\t2757196\t2941008\n",
      "    17\tMTB_anc\t2941009\t3124821\n",
      "    18\tMTB_anc\t3124822\t3308634\n",
      "    19\tMTB_anc\t3308635\t3492447\n",
      "    20\tMTB_anc\t3492448\t3676260\n",
      "    21\tMTB_anc\t3676261\t3860073\n",
      "    22\tMTB_anc\t3860074\t4043886\n",
      "    23\tMTB_anc\t4043887\t4227699\n",
      "    24\tMTB_anc\t4227700\t4411532\n",
      "Variant calling with bcftools parallel jobs\n",
      "Running variant calling for sample: DRR034421\n",
      "Running variant calling for sample: DRR034421\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Lines   total/split/realigned/skipped:\t1458/14/127/0\n",
      "Indexing the final VCF file\n",
      "Running variant calling for sample: DRR130077\n",
      "Running variant calling for sample: DRR130077\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Lines   total/split/realigned/skipped:\t1543/17/137/0\n",
      "Indexing the final VCF file\n",
      "Running variant calling for sample: DRR130078\n",
      "Running variant calling for sample: DRR130078\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Lines   total/split/realigned/skipped:\t1609/29/140/0\n",
      "Indexing the final VCF file\n",
      "Running variant calling for sample: DRR130093\n",
      "Running variant calling for sample: DRR130093\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Lines   total/split/realigned/skipped:\t1596/47/134/0\n",
      "Indexing the final VCF file\n"
     ]
    }
   ],
   "source": [
    "variant_calling ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LoFreq + bcftools\n",
    "\n",
    "Due to its high sensitivity in detecting low-frequency variants, crucial for identifying minor variants, the LoFreq tool was also employed [6]. The process involved two main steps: post-processing and variant calling. In the post-processing step, the `lofreq_postprocessing` function performed indel realignment on duplicated BAM files to refine read alignments around indels, followed by indexing the realigned BAM files. Subsequently, the `LoFreq_calling` function used `lofreq call-parallel` to detect variants in the realigned BAM files, producing VCF files with annotated and high-confidence variant calls. \n",
    "\n",
    "Additionally, `bcftools` was employed to further refine the variant calls by applying stringent filters and incorporating comprehensive annotations to the VCF files generated. This combined approach generated VCF files with low-frequency variants annotated with information regarding excluded loci, mappability, lineage-specific SNPs, and epitope region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "lofreq_postprocessing() { \n",
    "    local sample_file=\"$1\" \n",
    "    local genome_reference=\"$HOME/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta\"\n",
    "    local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\" \n",
    "    while IFS= read -r sample || [[ -n $sample ]]; do \n",
    "        if [ ! -f \"$sra_diretoria/$sample/${sample}_markdup.bam\" ]; then \n",
    "            echo \"No markdup.bam found for: $sample.\" \n",
    "            continue \n",
    "        fi \n",
    "        lofreq indelqual --dindel -f $genome_reference -o \"$sra_diretoria/$sample/${sample}_realigned.bam\" \"$sra_diretoria/$sample/${sample}_markdup.bam\"\n",
    "        samtools index \"$sra_diretoria/$sample/${sample}_realigned.bam\"\n",
    "    done < \"$sample_file\" \n",
    "    echo 'Post-processing with LoFreq completed.'\n",
    "}\n",
    "\n",
    "LoFreq_calling() {\n",
    "    local sample_file=\"$1\" \n",
    "    local genome_reference=\"$HOME/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta\"\n",
    "    local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\" \n",
    "    while IFS= read -r sample || [[ -n $sample ]]; do \n",
    "        if [ ! -f \"$sra_diretoria/$sample/${sample}_realigned.bam\" ]; then \n",
    "            echo \"No realigned.bam found for: $sample.\" \n",
    "            continue \n",
    "        fi \n",
    "        echo \"Performing variant calling for $sample\"\n",
    "        lofreq call-parallel --pp-threads 4 -f $genome_reference -o \"$sra_diretoria/$sample/${sample}_lofreq.vcf\" \"$sra_diretoria/$sample/${sample}_realigned.bam\" \n",
    "    done < \"$sample_file\"  \n",
    "    echo 'Variant calling using LoFreq completed.'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Post-processing with LoFreq completed.\n"
     ]
    }
   ],
   "source": [
    "lofreq_postprocessing ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing variant calling for DRR034421\n",
      "INFO [2024-06-16 22:44:16,019]: Using 4 threads with following basic args: lofreq call -f /home/ritanobrega00/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_realigned.bam\n",
      "\n",
      "INFO [2024-06-16 22:44:16,028]: Adding 12 commands to mp-pool\n",
      "Number of substitution tests performed: 619218\n",
      "Number of indel tests performed: 0\n",
      "INFO [2024-06-16 22:49:51,060]: Executing lofreq filter -i /tmp/lofreq2_call_parallel3c_gxp68/concat.vcf.gz -o /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_lofreq.vcf --snvqual-thresh 78 --indelqual-thresh 20\n",
      "\n",
      "Performing variant calling for DRR130077\n",
      "INFO [2024-06-16 22:49:51,174]: Using 4 threads with following basic args: lofreq call -f /home/ritanobrega00/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_realigned.bam\n",
      "\n",
      "INFO [2024-06-16 22:49:51,184]: Adding 12 commands to mp-pool\n",
      "Number of substitution tests performed: 573042\n",
      "Number of indel tests performed: 0\n",
      "INFO [2024-06-16 22:56:59,525]: Executing lofreq filter -i /tmp/lofreq2_call_parallelkc6lz639/concat.vcf.gz -o /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_lofreq.vcf --snvqual-thresh 78 --indelqual-thresh 20\n",
      "\n",
      "Performing variant calling for DRR130078\n",
      "INFO [2024-06-16 22:56:59,740]: Using 4 threads with following basic args: lofreq call -f /home/ritanobrega00/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_realigned.bam\n",
      "\n",
      "INFO [2024-06-16 22:56:59,758]: Adding 12 commands to mp-pool\n",
      "Number of substitution tests performed: 505410\n",
      "Number of indel tests performed: 0\n",
      "INFO [2024-06-16 23:11:07,982]: Executing lofreq filter -i /tmp/lofreq2_call_parallelnvgxr7zi/concat.vcf.gz -o /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_lofreq.vcf --snvqual-thresh 77 --indelqual-thresh 20\n",
      "\n",
      "Performing variant calling for DRR130093\n",
      "INFO [2024-06-16 23:11:08,185]: Using 4 threads with following basic args: lofreq call -f /home/ritanobrega00/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_realigned.bam\n",
      "\n",
      "INFO [2024-06-16 23:11:08,195]: Adding 12 commands to mp-pool\n",
      "Number of substitution tests performed: 972714\n",
      "Number of indel tests performed: 0\n",
      "INFO [2024-06-16 23:33:05,247]: Executing lofreq filter -i /tmp/lofreq2_call_parallelji0yhzsq/concat.vcf.gz -o /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_lofreq.vcf --snvqual-thresh 80 --indelqual-thresh 20\n",
      "\n",
      "Variant calling using LoFreq completed.\n"
     ]
    }
   ],
   "source": [
    "LoFreq_calling ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LoFreq + Bcftools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "lofreq_bcftools() {\n",
    "    local sample_file=\"$1\"\n",
    "    local genome_reference=\"$HOME/Projeto_Mtb/NGS_helper_files/MTB_anc.fasta\"\n",
    "    local sra_directory=\"$HOME/Projeto_Mtb/Sample_reads\"\n",
    "\n",
    "    echo \"Variant calling using LoFreq and bcftools\"\n",
    "    while IFS= read -r sample || [[ -n $sample ]]; do\n",
    "        local sample_dir=\"$sra_directory/$sample\"\n",
    "        local lofreq_vcf=\"$sample_dir/${sample}_lofreq.vcf\"\n",
    "        local bcftools_output=\"$sample_dir/${sample}_bcftools_lofreq.vcf.gz\"\n",
    "\n",
    "        if [ -s \"$lofreq_vcf\" ]; then\n",
    "            echo \"Running bcftools processing for sample: $sample\"\n",
    "            bgzip -c \"$sample_dir/${sample}_lofreq.vcf\" > \"$sample_dir/${sample}_lofreq.vcf.gz\"\n",
    "            tabix -p vcf \"$sample_dir/${sample}_lofreq.vcf.gz\"\n",
    "            bcftools annotate --annotations ~/Projeto_Mtb/NGS_helper_files/excludedloci_RLC2021_annot.tab.gz \\\n",
    "                             --header-lines ~/Projeto_Mtb/NGS_helper_files/excludedloci_RLC2021_annot.header \\\n",
    "                             --columns CHROM,FROM,TO,RLC_tag \\\n",
    "                             --threads \"$(nproc)\" --output-type u \"$lofreq_vcf.gz\" \\\n",
    "            | bcftools annotate --annotations ~/Projeto_Mtb/NGS_helper_files/blindspots_mappability_marin2021_annot.tab.gz \\\n",
    "                                --header-lines ~/Projeto_Mtb/NGS_helper_files/blindspots_mappability_marin2021_annot.header \\\n",
    "                                --columns CHROM,FROM,TO,Mappability \\\n",
    "                                --threads \"$(nproc)\" --output-type u \\\n",
    "            | bcftools annotate --annotations ~/Projeto_Mtb/NGS_helper_files/lineagesnps_annot.tab.gz \\\n",
    "                                --header-lines ~/Projeto_Mtb/NGS_helper_files/lineagesnps_annot.header \\\n",
    "                                --columns CHROM,POS,REF,ALT,Lineage_tag \\\n",
    "                                --threads \"$(nproc)\" --output-type u \\\n",
    "            | bcftools annotate --annotations ~/Projeto_Mtb/NGS_helper_files/iedbepitopes_annot.tab.gz \\\n",
    "                                --header-lines ~/Projeto_Mtb/NGS_helper_files/iedbepitopes_annot.header \\\n",
    "                                --columns CHROM,POS,REF,ALT,IEDB_tag \\\n",
    "                                --merge-logic IEDB_tag:unique \\\n",
    "                                --threads \"$(nproc)\" --output-type v \\\n",
    "            | bcftools annotate --set-id +'%CHROM:%POS' \\\n",
    "            | bgzip > \"$bcftools_output\"\n",
    "\n",
    "            if [ -s \"$bcftools_output\" ]; then\n",
    "                echo \"Indexing the final annotated VCF file for sample: $sample\"\n",
    "                tabix -p vcf \"$bcftools_output\"\n",
    "            else\n",
    "                echo \"Error: Failed to create the final annotated VCF file for sample: $sample\"\n",
    "            fi\n",
    "        else\n",
    "            echo \"Error: Failed to generate LoFreq VCF for sample: $sample\"\n",
    "        fi\n",
    "    done < \"$sample_file\"\n",
    "\n",
    "    echo \"Variant calling using LoFreq and bcftools completed.\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variant calling using LoFreq and bcftools\n",
      "Running bcftools processing for sample: DRR034421\n",
      "Indexing the final annotated VCF file for sample: DRR034421\n",
      "Running bcftools processing for sample: DRR130077\n",
      "Indexing the final annotated VCF file for sample: DRR130077\n",
      "Running bcftools processing for sample: DRR130078\n",
      "Indexing the final annotated VCF file for sample: DRR130078\n",
      "Running bcftools processing for sample: DRR130093\n",
      "Indexing the final annotated VCF file for sample: DRR130093\n",
      "Variant calling using LoFreq and bcftools completed.\n"
     ]
    }
   ],
   "source": [
    "lofreq_bcftools ~/Projeto_Mtb/list_isolates_SRA.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare VCF files from different variant callers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[tabix] the index file exists. Please use '-f' to overwrite.\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "tabix -p vcf \"$HOME/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_lofreq.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variants identified in DRR130078_simple.vcf:\n",
      "1576\n",
      "Variants identified in DRR130078_bcftools_varsonly_annotated.vcf:\n",
      "1640\n",
      "Variants identified in DRR130078_lofreq.vcf:\n",
      "1743\n"
     ]
    }
   ],
   "source": [
    "vcf_files=(\"DRR130078_simple.vcf\" \"DRR130078_bcftools_varsonly_annotated.vcf\" \"DRR130078_lofreq.vcf\")\n",
    "for vcf_file in \"${vcf_files[@]}\"\n",
    "do\n",
    "    echo \"Variants identified in $vcf_file:\"\n",
    "    grep -c \"^MTB_anc\" $HOME/Projeto_Mtb/Sample_reads/DRR130078/\"$vcf_file\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variants identified in: DRR130077_simple.vcf\n",
      "1532\n",
      "Variants identified in: DRR130077_bcftools_varsonly_annotated.vcf\n",
      "1560\n",
      "Variants identified in: DRR130077_lofreq.vcf\n",
      "1616\n"
     ]
    }
   ],
   "source": [
    "vcf_files=(\"DRR130077_simple.vcf\" \"DRR130077_bcftools_varsonly_annotated.vcf\" \"DRR130077_lofreq.vcf\")\n",
    "for vcf_file in \"${vcf_files[@]}\"\n",
    "do\n",
    "    echo \"Variants identified in $vcf_file:\"\n",
    "    grep -c \"^MTB_anc\" $HOME/Projeto_Mtb/Sample_reads/DRR130077/\"$vcf_file\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variants identified in: DRR034421_simple.vcf\n",
      "1417\n",
      "Variants identified in: DRR034421_bcftools_varsonly_annotated.vcf\n",
      "1472\n",
      "Variants identified in: DRR034421_lofreq.vcf\n",
      "1346\n"
     ]
    }
   ],
   "source": [
    "vcf_files=(\"DRR034421_simple.vcf\" \"DRR034421_bcftools_varsonly_annotated.vcf\" \"DRR034421_lofreq.vcf\")\n",
    "for vcf_file in \"${vcf_files[@]}\"\n",
    "do\n",
    "    echo \"Variants identified in $vcf_file:\"\n",
    "    grep -c \"^MTB_anc\" $HOME/Projeto_Mtb/Sample_reads/DRR034421/\"$vcf_file\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variants identified in DRR130093_simple.vcf:\n",
      "1571\n",
      "Variants identified in DRR130093_bcftools_varsonly_annotated.vcf:\n",
      "1645\n",
      "Variants identified in DRR130093_lofreq.vcf:\n",
      "1723\n"
     ]
    }
   ],
   "source": [
    "vcf_files=(\"DRR130093_simple.vcf\" \"DRR130093_bcftools_varsonly_annotated.vcf\" \"DRR130093_lofreq.vcf.gz\")\n",
    "for vcf_file in \"${vcf_files[@]}\"\n",
    "do\n",
    "    echo \"Variants identified in $vcf_file:\"\n",
    "    \n",
    "    grep -c \"^MTB_anc\" $HOME/Projeto_Mtb/Sample_reads/DRR130093/\"$vcf_file\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "comparing_VCF(){\n",
    "    local sample=\"$1\" \n",
    "    local sra_diretoria=\"$HOME/Projeto_Mtb/Sample_reads\" \n",
    "\n",
    "    echo \"Comparing VCF files for $sample:\"\n",
    "    bgzip -c \"$sra_diretoria/${sample}/${sample}_simple.vcf\" > \"$sra_diretoria/${sample}/${sample}_simple.vcf.gz\"\n",
    "    bgzip -c \"$sra_diretoria/${sample}/${sample}_bcftools_varsonly_annotated.vcf\" > \"$sra_diretoria/${sample}/${sample}_bcftools_varsonly_annotated.vcf.gz\"\n",
    "\n",
    "    tabix -p vcf \"$sra_diretoria/${sample}/${sample}_simple.vcf.gz\"\n",
    "    tabix -p vcf \"$sra_diretoria/${sample}/${sample}_bcftools_varsonly_annotated.vcf.gz\"\n",
    "    tabix -p vcf \"$sra_diretoria/${sample}/${sample}_bcftools_lofreq.vcf.gz\"\n",
    "    \n",
    "    vcf-compare \"$sra_diretoria/${sample}/${sample}_simple.vcf.gz\" \"$sra_diretoria/${sample}/${sample}_bcftools_varsonly_annotated.vcf.gz\" \"$sra_diretoria/${sample}/${sample}_bcftools_lofreq.vcf.gz\" \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing VCF files for DRR130078:\n",
      "[tabix] the index file exists. Please use '-f' to overwrite.\n",
      "# This file was generated by vcf-compare.\n",
      "# The command line was: vcf-compare(v0.1.14-12-gcdb80b8) /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_simple.vcf.gz /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_varsonly_annotated.vcf.gz /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_lofreq.vcf.gz\n",
      "#\n",
      "#VN 'Venn-Diagram Numbers'. Use `grep ^VN | cut -f 2-` to extract this part.\n",
      "#VN The columns are: \n",
      "#VN        1  .. number of sites unique to this particular combination of files\n",
      "#VN        2- .. combination of files and space-separated number, a fraction of sites in the file\n",
      "VN\t5\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_lofreq.vcf.gz (0.3%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_varsonly_annotated.vcf.gz (0.3%)\n",
      "VN\t19\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_simple.vcf.gz (1.2%)\n",
      "VN\t22\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_lofreq.vcf.gz (1.3%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_simple.vcf.gz (1.4%)\n",
      "VN\t69\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_varsonly_annotated.vcf.gz (4.3%)\n",
      "VN\t266\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_varsonly_annotated.vcf.gz (16.5%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_simple.vcf.gz (16.9%)\n",
      "VN\t447\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_lofreq.vcf.gz (25.6%)\n",
      "VN\t1269\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_lofreq.vcf.gz (72.8%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_varsonly_annotated.vcf.gz (78.9%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_simple.vcf.gz (80.5%)\n",
      "#SN Summary Numbers. Use `grep ^SN | cut -f 2-` to extract this part.\n",
      "SN\tNumber of REF matches:\t1269\n",
      "SN\tNumber of ALT matches:\t1253\n",
      "SN\tNumber of REF mismatches:\t0\n",
      "SN\tNumber of ALT mismatches:\t16\n",
      "SN\tNumber of samples in GT comparison:\t0\n",
      "# Number of sites lost due to grouping (e.g. duplicate sites): lost, %lost, read, reported, file\n",
      "SN\tNumber of lost sites:\t31\t1.9%\t1640\t1609\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130078/DRR130078_bcftools_varsonly_annotated.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "comparing_VCF DRR130078"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing VCF files for DRR130077:\n",
      "# This file was generated by vcf-compare.\n",
      "# The command line was: vcf-compare(v0.1.14-12-gcdb80b8) /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_simple.vcf.gz /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_bcftools_varsonly_annotated.vcf.gz /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_lofreq.vcf.gz\n",
      "#\n",
      "#VN 'Venn-Diagram Numbers'. Use `grep ^VN | cut -f 2-` to extract this part.\n",
      "#VN The columns are: \n",
      "#VN        1  .. number of sites unique to this particular combination of files\n",
      "#VN        2- .. combination of files and space-separated number, a fraction of sites in the file\n",
      "VN\t3\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_bcftools_varsonly_annotated.vcf.gz (0.2%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_lofreq.vcf.gz (0.2%)\n",
      "VN\t21\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_simple.vcf.gz (1.4%)\n",
      "VN\t24\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_lofreq.vcf.gz (1.5%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_simple.vcf.gz (1.6%)\n",
      "VN\t53\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_bcftools_varsonly_annotated.vcf.gz (3.4%)\n",
      "VN\t234\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_bcftools_varsonly_annotated.vcf.gz (15.2%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_simple.vcf.gz (15.3%)\n",
      "VN\t336\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_lofreq.vcf.gz (20.8%)\n",
      "VN\t1253\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_bcftools_varsonly_annotated.vcf.gz (81.2%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_lofreq.vcf.gz (77.5%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_simple.vcf.gz (81.8%)\n",
      "#SN Summary Numbers. Use `grep ^SN | cut -f 2-` to extract this part.\n",
      "SN\tNumber of REF matches:\t1252\n",
      "SN\tNumber of ALT matches:\t1241\n",
      "SN\tNumber of REF mismatches:\t1\n",
      "SN\tNumber of ALT mismatches:\t11\n",
      "SN\tNumber of samples in GT comparison:\t0\n",
      "# Number of sites lost due to grouping (e.g. duplicate sites): lost, %lost, read, reported, file\n",
      "SN\tNumber of lost sites:\t17\t1.1%\t1560\t1543\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130077/DRR130077_bcftools_varsonly_annotated.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "comparing_VCF DRR130077"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing VCF files for DRR034421:\n",
      "[tabix] the index file exists. Please use '-f' to overwrite.\n",
      "# This file was generated by vcf-compare.\n",
      "# The command line was: vcf-compare(v0.1.14-12-gcdb80b8) /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_simple.vcf.gz /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_varsonly_annotated.vcf.gz /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_lofreq.vcf.gz\n",
      "#\n",
      "#VN 'Venn-Diagram Numbers'. Use `grep ^VN | cut -f 2-` to extract this part.\n",
      "#VN The columns are: \n",
      "#VN        1  .. number of sites unique to this particular combination of files\n",
      "#VN        2- .. combination of files and space-separated number, a fraction of sites in the file\n",
      "VN\t4\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_lofreq.vcf.gz (0.3%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_varsonly_annotated.vcf.gz (0.3%)\n",
      "VN\t4\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_lofreq.vcf.gz (0.3%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_simple.vcf.gz (0.3%)\n",
      "VN\t16\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_simple.vcf.gz (1.1%)\n",
      "VN\t57\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_varsonly_annotated.vcf.gz (3.9%)\n",
      "VN\t220\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_lofreq.vcf.gz (16.3%)\n",
      "VN\t279\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_varsonly_annotated.vcf.gz (19.1%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_simple.vcf.gz (19.7%)\n",
      "VN\t1118\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_lofreq.vcf.gz (83.1%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_varsonly_annotated.vcf.gz (76.7%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_simple.vcf.gz (78.9%)\n",
      "#SN Summary Numbers. Use `grep ^SN | cut -f 2-` to extract this part.\n",
      "SN\tNumber of REF matches:\t1118\n",
      "SN\tNumber of ALT matches:\t1108\n",
      "SN\tNumber of REF mismatches:\t0\n",
      "SN\tNumber of ALT mismatches:\t10\n",
      "SN\tNumber of samples in GT comparison:\t0\n",
      "# Number of sites lost due to grouping (e.g. duplicate sites): lost, %lost, read, reported, file\n",
      "SN\tNumber of lost sites:\t14\t1.0%\t1472\t1458\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR034421/DRR034421_bcftools_varsonly_annotated.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "comparing_VCF DRR034421"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing VCF files for DRR130093:\n",
      "[tabix] the index file exists. Please use '-f' to overwrite.\n",
      "# This file was generated by vcf-compare.\n",
      "# The command line was: vcf-compare(v0.1.14-12-gcdb80b8) /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_simple.vcf.gz /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_varsonly_annotated.vcf.gz /home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_lofreq.vcf.gz\n",
      "#\n",
      "#VN 'Venn-Diagram Numbers'. Use `grep ^VN | cut -f 2-` to extract this part.\n",
      "#VN The columns are: \n",
      "#VN        1  .. number of sites unique to this particular combination of files\n",
      "#VN        2- .. combination of files and space-separated number, a fraction of sites in the file\n",
      "VN\t3\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_lofreq.vcf.gz (0.2%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_varsonly_annotated.vcf.gz (0.2%)\n",
      "VN\t20\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_lofreq.vcf.gz (1.2%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_simple.vcf.gz (1.3%)\n",
      "VN\t28\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_simple.vcf.gz (1.8%)\n",
      "VN\t70\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_varsonly_annotated.vcf.gz (4.4%)\n",
      "VN\t247\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_varsonly_annotated.vcf.gz (15.5%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_simple.vcf.gz (15.7%)\n",
      "VN\t424\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_lofreq.vcf.gz (24.6%)\n",
      "VN\t1276\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_lofreq.vcf.gz (74.1%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_varsonly_annotated.vcf.gz (79.9%)\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_simple.vcf.gz (81.2%)\n",
      "#SN Summary Numbers. Use `grep ^SN | cut -f 2-` to extract this part.\n",
      "SN\tNumber of REF matches:\t1276\n",
      "SN\tNumber of ALT matches:\t1239\n",
      "SN\tNumber of REF mismatches:\t0\n",
      "SN\tNumber of ALT mismatches:\t37\n",
      "SN\tNumber of samples in GT comparison:\t0\n",
      "# Number of sites lost due to grouping (e.g. duplicate sites): lost, %lost, read, reported, file\n",
      "SN\tNumber of lost sites:\t49\t3.0%\t1645\t1596\t/home/ritanobrega00/Projeto_Mtb/Sample_reads/DRR130093/DRR130093_bcftools_varsonly_annotated.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "comparing_VCF DRR130093"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References:\n",
    "\n",
    "[1] https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump\n",
    "\n",
    "[2] https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/\n",
    "\n",
    "[3]  https://github.com/lh3/bwa\n",
    "\n",
    "[4] https://github.com/brentp/mosdepth\n",
    "\n",
    "[5] https://samtools.github.io/bcftools/bcftools.html\n",
    "\n",
    "[6] Wilm, A., Aw, P. P., Bertrand, D., Yeo, G. H., Ong, S. H., Wong, C. H., Khor, C. C., Petric, R., Hibberd, M. L., & Nagarajan, N. (2012). LoFreq: a sequence-quality aware, ultra-sensitive variant caller for uncovering cell-population heterogeneity from high-throughput sequencing datasets. Nucleic acids research, 40(22), 11189–11201. https://doi.org/10.1093/nar/gks918\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
